### ISSUE TITLE
UserCallbackException is thrown when another Exception is expected

### ISSUE THREAD

**Issue by waldfee** [FakeItEasy/FakeItEasy#1543]
I upgraded FakeItEasy from 2.3.1 to 4.9.1 and I have the following issue.

**Scenario:**
1. A Fake is configured to throw
2. Another Fake is configured to indirectly call the throwing fake
3. The exception from the first Fake is catched
4. Another exception is thrown in the catch block

**Expected behaviour:**
The exception thrown in the catch block is raised to the caller

**Observed behaviour:**
An UserCallbackException is raised to the caller

**Workarounds:**
- Unpack the UserCallbackException and assert it's InnerException
- Use a manually written Fake for the throwing interface

**This behaviour seems to have been introduced with #1364
Is this behaviour intended, and is there a way to configure it, or do I have to use one of the workarounds?**

**FakeItEasy version**: 4.9.1, also happens in 5.0.0-alpha.1
.NET Framework version: 4.7.2

Minimal repro:
```c#
[TestClass]
public class UserCallbackExceptionTest
{
	public class CatchedException : Exception
	{
	}

	public class ThrowedException : Exception
	{
	}

	public interface IThrows
	{
		void ThrowIt();
	}

	public interface IDoStuff
	{
		void Execute(Action operation);
	}

	public class UnderTest
	{
		private readonly IThrows _throws;
		private readonly IDoStuff _dostuff;

		public UnderTest(IThrows throws, IDoStuff dostuff)
		{
			_throws = throws;
			_dostuff = dostuff;
		}

		public void DoStuff()
		{
			_dostuff.Execute(DoStuffPrivately);
		}

		private void DoStuffPrivately()
		{
			try
			{
				_throws.ThrowIt();
			}
			catch (CatchedException)
			{
				throw new ThrowedException();
			}
		}
	}

	public class HandWrittenFakeDoStuff : IDoStuff
	{
		public void Execute(Action operation)
		{
			operation();
		}
	}

	[TestMethod]
	public void WorksWhenUsingHandWrittenFake()
	{
		IThrows throws = A.Fake<IThrows>();
		A.CallTo(() => throws.ThrowIt())
			.Throws(new CatchedException());

		UnderTest underTest = new UnderTest(throws, new HandWrittenFakeDoStuff());

		Assert.ThrowsException<ThrowedException>(() => underTest.DoStuff());
	}

	[TestMethod]
	public void DoesNotWorkWhenUsingFakeItEasyFake()
	{
		IDoStuff doStuff = A.Fake<IDoStuff>();
		A.CallTo(() => doStuff.Execute(A<Action>.Ignored))
			.Invokes(call =>
					 {
						 // assume the real implementation does other things beside calling the action
						 // to avoid those sideeffects we fake it
						 Action operation = call.GetArgument<Action>(0);
						 operation();
					 });

		IThrows throws = A.Fake<IThrows>();
		A.CallTo(() => throws.ThrowIt())
			.Throws(new CatchedException());

		UnderTest underTest = new UnderTest(throws, doStuff);

		Assert.ThrowsException<ThrowedException>(() => underTest.DoStuff());
	}
}
```

**Comment by thomaslevesque** [comment]
Hi @waldfee,

Thanks for reporting this. At first glance, it looks like it's working as designed, but I see how this can be a problem in your case. Basically, starting with #1364, we wrap exceptions that occur in user callbacks, but in retrospect it might not be the right thing to do for `Invokes`. The callback specified in `Invokes` actually stands for the body of the faked method, so it could legitimately do anything the method body could do, including throwing an exception.

We'll  fix this ASAP.

**Comment by blairconrad** [comment]
Oh, and @waldfee, great explanation, reproduction, and assessment.

**Comment by thomaslevesque** [comment]
This fix has been released as part of [FakeItEasy 4.9.2](https://github.com/FakeItEasy/FakeItEasy/releases/4.9.2).

Thanks, @waldfee! The reproduction instructions were excellent. Look for your name in the release notes. :trophy:
